File: apps/macos/Doctor.swift
// Copyright 2026 Shazron Abdullah and Bunyi contributors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//
//  Doctor.swift
//  Bunyi
//
//  Preflight checks: can this machine finish a generation right now?
//

import Foundation
import MLX

/// How much a finding matters.
///
/// The distinction that earns its keep is `blocker` versus `warning`. A blocker
/// is a fact about now — there is no room on the disk, the server does not
/// answer — and acting on it saves the user a wait that could not have
/// succeeded. A warning is a prediction, and the app refuses to stop a run on
/// one, because a prediction that stops work is wrong in the expensive
/// direction: it refuses runs that would have finished.
enum DoctorSeverity {
    case ok
    case warning
    case blocker
}

/// One check's result.
struct DoctorFinding: Identifiable {
    let id = UUID()
    /// What was checked, as a noun phrase: "Disk space", "Memory".
    let title: String
    /// What was found, and what to do about it. Per `spec/FEATURES.md` §10 this
    /// is the actionable half — "Free up 4.2 GB", not "insufficient space".
    let detail: String
    let severity: DoctorSeverity

    var symbolName: String {
        switch severity {
        case .ok:      "checkmark.circle.fill"
        case .warning: "exclamationmark.triangle.fill"
        case .blocker: "xmark.octagon.fill"
        }
    }
}

/// Everything a run found.
struct DoctorReport {
    /// Which mode's model this is about.
    ///
    /// Carried rather than inferred because it is not always on screen. The
    /// History tab has no mode of its own, so a run started there reports on
    /// the last one generated with — perfectly sensible, and completely opaque
    /// unless the report says so.
    let mode: TTSMode
    let findings: [DoctorFinding]

    var blockers: [DoctorFinding] { findings.filter { $0.severity == .blocker } }
    var warnings: [DoctorFinding] { findings.filter { $0.severity == .warning } }

    /// Nothing stands in the way of a generation. Warnings do not count — see
    /// `DoctorSeverity`.
    var isClear: Bool { blockers.isEmpty }

    /// One line per finding, for the Logs. Deliberately flat text: the Logs
    /// window is what gets pasted into a bug report, and a report is only
    /// useful there if it survives being copied as plain text.
    var logLines: [String] {
        ["Doctor: checking \(mode.rawValue)"] + findings.map { finding in
            let mark = switch finding.severity {
            case .ok:      "ok"
            case .warning: "warning"
            case .blocker: "blocked"
            }
            return "Doctor: [\(mark)] \(finding.title) — \(finding.detail)"
        }
    }
}

/// The checks themselves.
///
/// Deliberately not an `@Observable` model: a run is a function of the machine
/// at one instant, not state the app carries around. Whoever asked holds the
/// report for as long as they are showing it.
@MainActor
enum Doctor {

    /// How much room beyond the model's own size counts as comfortable.
    ///
    /// A download that exactly fits leaves a full disk behind it, and the next
    /// thing to write anything — including this app, saving the WAV it just
    /// generated — fails instead. Below this is a warning, not a refusal.
    private static let comfortableDiskHeadroom: Int64 = 5_000_000_000

    /// What a generation needs on top of the weights: activations, the audio
    /// being assembled, and the codec. A fraction rather than a constant
    /// because it scales with the model — the 1.7B models want proportionally
    /// more working room than the 0.6B one.
    private static let memoryWorkingFactor = 1.3

    /// Space to keep free for output, when no download is pending. Generated
    /// clips are megabytes, so this is about not writing onto a disk that is
    /// already at the edge rather than about the file itself.
    private static let outputHeadroom: Int64 = 500_000_000

    /// Run the checks.
    ///
    /// - Parameter deep: also verify downloaded files against the server's
    ///   published digests. Hashing gigabytes takes real time, so this is for
    ///   the on-demand run only — never the one before a generation.
    static func run(mode: TTSMode,
                    engine: TTSEngine,
                    deep: Bool = false) async -> DoctorReport {
        var findings: [DoctorFinding] = []

        let modelDir = TTSEngine.modelDirectory(for: mode)
        let isPresent = TTSEngine.isModelComplete(for: mode)
        let modelsRoot = ModelsLocation.current()

        findings.append(modelFinding(mode: mode, isPresent: isPresent))

        // Sizing the model folder walks it, and hashing reads all of it. Both
        // are disk work and neither belongs on the actor driving the window.
        let modelBytes: Int64 = isPresent
            ? await Task.detached(priority: .userInitiated) {
                ModelStore.size(of: modelDir)
            }.value
            : Int64(mode.approxDownloadBytes)

        findings.append(diskFinding(modelsRoot: modelsRoot,
                                    needsDownload: !isPresent,
                                    modelBytes: modelBytes))
        findings.append(memoryFinding(mode: mode,
                                      modelBytes: modelBytes,
                                      isPresent: isPresent))

        // Only worth asking when a run would actually reach for the network. A
        // reachability failure on a model already on disk is not a problem the
        // user has — the app is offline-capable once the files are there.
        if !isPresent {
            findings.append(await sourceFinding(mode: mode))
        }

        findings.append(outputFinding(folder: engine.outputsFolder))

        if deep, isPresent {
            findings.append(await integrityFinding(mode: mode,
                                                   engine: engine,
                                                   dir: modelDir))
        }

        return DoctorReport(mode: mode, findings: findings)
    }

    // MARK: - Individual checks

    private static func modelFinding(mode: TTSMode, isPresent: Bool) -> DoctorFinding {
        // A missing model is not a fault. It is the first-run state, and the
        // app's answer to it is to download one. Reporting it as a failure
        // would make a healthy first launch look broken; what it actually does
        // is change what the disk and network checks are about.
        guard !isPresent else {
            return DoctorFinding(
                title: "Model",
                detail: "\(mode.rawValue) is downloaded and complete.",
                severity: .ok)
        }
        let size = Int64(mode.approxDownloadBytes)
            .formatted(.byteCount(style: .file))
        return DoctorFinding(
            title: "Model",
            detail: """
                \(mode.rawValue) is not downloaded yet. Generating will \
                download it first, about \(size).
                """,
            severity: .ok)
    }

    private static func diskFinding(modelsRoot: URL,
                                    needsDownload: Bool,
                                    modelBytes: Int64) -> DoctorFinding {
        // Measured on the volume holding the models folder, which Settings
        // lets the user move to another disk entirely. Checking the boot
        // volume would answer a question nobody asked.
        guard let free = freeBytes(at: modelsRoot) else {
            return DoctorFinding(
                title: "Disk space",
                detail: """
                    Could not read free space on the volume holding the models \
                    folder (\(modelsRoot.path)). Check that the disk is still \
                    connected.
                    """,
                severity: .warning)
        }

        let freeText = free.formatted(.byteCount(style: .file))
        guard needsDownload else {
            guard free < outputHeadroom else {
                return DoctorFinding(
                    title: "Disk space",
                    detail: "\(freeText) free. The model is already downloaded.",
                    severity: .ok)
            }
            return DoctorFinding(
                title: "Disk space",
                detail: """
                    Only \(freeText) free. There is room for the model, but \
                    little for the audio it produces.
                    """,
                severity: .warning)
        }

        let needText = modelBytes.formatted(.byteCount(style: .file))
        guard free >= modelBytes else {
            let short = (modelBytes - free).formatted(.byteCount(style: .file))
            return DoctorFinding(
                title: "Disk space",
                detail: """
                    \(freeText) free, and the download needs about \(needText). \
                    Free up \(short) or choose another models folder in \
                    Settings → Storage.
                    """,
                severity: .blocker)
        }
        guard free >= modelBytes + comfortableDiskHeadroom else {
            return DoctorFinding(
                title: "Disk space",
                detail: """
                    \(freeText) free and the download needs about \(needText), \
                    which fits but leaves the disk nearly full.
                    """,
                severity: .warning)
        }
        return DoctorFinding(
            title: "Disk space",
            detail: "\(freeText) free, and the download needs about \(needText).",
            severity: .ok)
    }

    private static func memoryFinding(mode: TTSMode,
                                      modelBytes: Int64,
                                      isPresent: Bool) -> DoctorFinding {
        let needed = Int64(Double(modelBytes) * memoryWorkingFactor)
        let available = availableMemory()
        let neededText = needed.formatted(.byteCount(style: .memory))
        let availableText = available.formatted(.byteCount(style: .memory))

        // Never a blocker. The number moves the instant another app quits, and
        // a machine short of memory still finishes — it swaps, which is slow
        // rather than fatal. Refusing here would refuse runs that work.
        guard available < needed else {
            return DoctorFinding(
                title: "Memory",
                detail: """
                    \(availableText) available, and the \(mode.rawValue) model \
                    wants about \(neededText).
                    """,
                severity: .ok)
        }
        let unknownSize = isPresent ? "" : " once it has downloaded"
        return DoctorFinding(
            title: "Memory",
            detail: """
                \(availableText) available and the \(mode.rawValue) model wants \
                about \(neededText)\(unknownSize). It will still run, but the Mac may \
                swap, which can make generation slow and playback stutter. \
                Quitting other apps first will help.
                """,
            severity: .warning)
    }

    private static func sourceFinding(mode: TTSMode) async -> DoctorFinding {
        let probe: URL
        let where_: String
        switch mode.effectiveSource {
        case .repo(let repoID):
            guard let url = URL(string:
                "https://huggingface.co/\(repoID)/resolve/main/config.json") else {
                return DoctorFinding(
                    title: "Model source",
                    detail: "\"\(repoID)\" is not a usable repository name.",
                    severity: .blocker)
            }
            probe = url
            where_ = "Hugging Face"
        case .baseURL(let base):
            probe = base.appendingPathComponent("config.json")
            where_ = base.host ?? base.absoluteString
        }

        guard await isReachable(probe) else {
            // Named as a network problem, not a model problem. Told "model
            // missing" when their own server is down, someone goes looking in
            // the wrong place — and self-hosting is exactly where this check
            // pays for itself.
            return DoctorFinding(
                title: "Model source",
                detail: """
                    \(where_) did not answer, so the download cannot start. \
                    Check your connection, or the source for this mode in \
                    Settings → Models.
                    """,
                severity: .blocker)
        }
        return DoctorFinding(
            title: "Model source",
            detail: "\(where_) is reachable.",
            severity: .ok)
    }

    private static func outputFinding(folder: URL) -> DoctorFinding {
        let fm = FileManager.default
        try? fm.createDirectory(at: folder, withIntermediateDirectories: true)
        guard fm.isWritableFile(atPath: folder.path) else {
            return DoctorFinding(
                title: "Output folder",
                detail: """
                    Bunyi cannot write to \(folder.path), so a finished \
                    generation could not be saved.
                    """,
                severity: .blocker)
        }
        return DoctorFinding(
            title: "Output folder",
            detail: "Generated audio can be written to \(folder.path).",
            severity: .ok)
    }

    private static func integrityFinding(mode: TTSMode,
                                         engine: TTSEngine,
                                         dir: URL) async -> DoctorFinding {
        guard let entries = await engine.publishedDigests(for: mode) else {
            return DoctorFinding(
                title: "Model files",
                detail: """
                    This model's source publishes no checksums, so the files \
                    cannot be verified. Nothing is wrong — there is just \
                    nothing to check against.
                    """,
                severity: .ok)
        }
        let withDigests = entries.filter { $0.sha256 != nil }
        guard !withDigests.isEmpty else {
            return DoctorFinding(
                title: "Model files",
                detail: "The published manifest lists files but no checksums.",
                severity: .ok)
        }

        let bad = await Task.detached(priority: .utility) {
            withDigests.filter { entry in
                let file = dir.appendingPathComponent(entry.path)
                // The downloader's own hash, not a second one. A file is
                // "intact" if it would satisfy the check the download path
                // applies, and two implementations of that could disagree.
                guard let actual = try? HTTPFileDownloader.sha256Hex(of: file)
                else { return true }
                return actual != entry.sha256
            }.map(\.path)
        }.value

        guard bad.isEmpty else {
            // This is the failure that otherwise loads and speaks nonsense, so
            // it names the files: "the model is corrupt" leaves nothing to do,
            // whereas a named file can be deleted and fetched again.
            let names = bad.prefix(3).joined(separator: ", ")
            let more = bad.count > 3 ? " and \(bad.count - 3) more" : ""
            return DoctorFinding(
                title: "Model files",
                detail: """
                    \(bad.count) file\(bad.count == 1 ? "" : "s") do not match \
                    the published checksums: \(names)\(more). Delete \(mode.rawValue) \
                    in Settings → Storage and download it again.
                    """,
                severity: .blocker)
        }
        return DoctorFinding(
            title: "Model files",
            detail: """
                All \(withDigests.count) files match the checksums the server \
                publishes.
                """,
            severity: .ok)
    }

    // MARK: - Machine facts

    private static func freeBytes(at url: URL) -> Int64? {
        // ForImportantUsage rather than the plain capacity: it counts space the
        // system would purge to satisfy a real write, which is what a download
        // about to start actually gets.
        let values = try? url.resourceValues(
            forKeys: [.volumeAvailableCapacityForImportantUsageKey])
        return values?.volumeAvailableCapacityForImportantUsage
    }

    /// Memory a large allocation could actually get.
    ///
    /// Not `physicalMemory` minus what is used: on macOS most of "used" is
    /// reclaimable. Free, inactive and purgeable pages are the ones the kernel
    /// will hand over without swapping, which is the question being asked.
    ///
    /// MLX's own buffer cache is added because the engine gives it back before
    /// a generation starts — counting it as unavailable would have Doctor warn
    /// about memory the app is about to release itself.
    private static func availableMemory() -> Int64 {
        var stats = vm_statistics64_data_t()
        var count = mach_msg_type_number_t(
            MemoryLayout<vm_statistics64_data_t>.size / MemoryLayout<integer_t>.size)
        let result = withUnsafeMutablePointer(to: &stats) {
            $0.withMemoryRebound(to: integer_t.self, capacity: Int(count)) {
                host_statistics64(mach_host_self(), HOST_VM_INFO64, $0, &count)
            }
        }
        guard result == KERN_SUCCESS else {
            // Unknown rather than zero: zero would fire the low-memory warning
            // on every run, and a warning that is always on is noise.
            return Int64(ProcessInfo.processInfo.physicalMemory)
        }
        // sysconf rather than `vm_kernel_page_size`: that is an imported
        // global var, which Swift 6 will not let a concurrent context read.
        let page = Int64(sysconf(_SC_PAGESIZE))
        let reclaimable = Int64(stats.free_count)
            + Int64(stats.inactive_count)
            + Int64(stats.purgeable_count)
        return reclaimable * page + Int64(GPU.cacheMemory)
    }

    private static func isReachable(_ url: URL) async -> Bool {
        var request = URLRequest(url: url)
        request.httpMethod = "HEAD"
        // Short: this runs before every generation on a first run, and a dead
        // host that takes sixty seconds to admit it would be worse than the
        // failure it is reporting.
        request.timeoutInterval = 10
        guard let (_, response) = try? await URLSession.shared.data(for: request),
              let http = response as? HTTPURLResponse else { return false }
        return (200..<400).contains(http.statusCode)
    }
}

File: apps/dotnet/src/Core/Diagnostics/Doctor.cs
// Copyright 2026 Shazron Abdullah and Bunyi contributors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System.Text;
using Bunyi.Core.Models;

using Bunyi.Core.Engine;

namespace Bunyi.Core.Diagnostics;

/// <summary>How bad a finding is (spec §11).</summary>
public enum DoctorSeverity
{
    Ok,
    Warning,
    Blocker,
}

/// <summary>One thing Doctor looked at.</summary>
/// <remarks>
/// Every finding names what is wrong <b>and what would resolve it</b> — §11 is
/// explicit that Doctor fixes nothing, so a finding that only reports a state
/// leaves the user with nothing to do.
/// </remarks>
public sealed record DoctorFinding(string Title, string Detail, DoctorSeverity Severity);

/// <summary>Everything Doctor found, for one mode.</summary>
public sealed record DoctorReport(TtsMode Mode, IReadOnlyList<DoctorFinding> Findings)
{
    /// <summary>Whether anything would stop a run.</summary>
    public bool HasBlockers => Findings.Any(f => f.Severity == DoctorSeverity.Blocker);

    /// <summary>Findings that would stop a run.</summary>
    public IEnumerable<DoctorFinding> Blockers =>
        Findings.Where(f => f.Severity == DoctorSeverity.Blocker);

    /// <summary>Findings worth saying but not worth stopping for.</summary>
    public IEnumerable<DoctorFinding> Warnings =>
        Findings.Where(f => f.Severity == DoctorSeverity.Warning);

    /// <summary>
    /// The report as text, for the Logs and the clipboard.
    /// </summary>
    /// <remarks>
    /// It names the mode. §11: the checks are per-mode — three modes, three
    /// models, three sizes, three separately configured sources — so "the
    /// model" is ambiguous unless it is said which. It matters most when
    /// there is no mode on screen at all, which is the case from History.
    /// </remarks>
    public string Describe()
    {
        var text = new StringBuilder();
        text.AppendLine($"Doctor — {Mode.DisplayName()}");

        foreach (var finding in Findings)
        {
            var mark = finding.Severity switch
            {
                DoctorSeverity.Blocker => "STOP",
                DoctorSeverity.Warning => "note",
                _ => "ok",
            };
            text.AppendLine($"  [{mark}] {finding.Title}: {finding.Detail}");
        }

        return text.ToString().TrimEnd();
    }
}

/// <summary>
/// Can this machine finish a generation right now? (spec §11)
/// </summary>
/// <remarks>
/// <para>
/// Not a settings panel, and it fixes nothing. It runs before every generation
/// — <b>before any download begins</b>, because the point is not to discover
/// after 3.4 GB that there was never room for it — and on demand from the
/// window.
/// </para>
/// <para>
/// A report is a value rather than carried state, so it is a static function
/// over what it was given. Nothing here mutates anything.
/// </para>
/// </remarks>
public static class Doctor
{
    /// <summary>
    /// How much larger than its files a model is once resident.
    /// </summary>
    /// <remarks>
    /// Measured, not guessed: the 0.6B preset export is 5.88 GB on disk and
    /// peaked at 8.73 GB resident, which is 1.48x. macOS uses 1.3 for MLX; this
    /// runtime holds the talker's prefill and decode graphs as separate
    /// sessions over the same weights, so it needs more. See RESEARCH-ONNX.md.
    /// </remarks>
    public const double MemoryWorkingFactor = 1.5;

    /// <summary>Headroom wanted beyond the download itself, before warning.</summary>
    public static readonly long DiskWarningHeadroom = 5L * 1000 * 1000 * 1000;

    /// <summary>
    /// Space a generation's own output needs, when nothing is downloading.
    /// </summary>
    /// <remarks>
    /// Clips are megabytes, so this is about not writing onto a disk already at
    /// the edge rather than about the file itself.
    /// </remarks>
    public static readonly long OutputHeadroom = 500L * 1000 * 1000;

    /// <summary>Runs the checks.</summary>
    /// <param name="deep">
    /// Also verify files against the digests the server published. Hashing
    /// gigabytes takes real time, so this is for the on-demand run only — never
    /// the one before a generation.
    /// </param>
    public static async Task<DoctorReport> RunAsync(
        TtsMode mode,
        ModelSource source,
        ModelLayout layout,
        string modelsRoot,
        string outputFolder,
        ISystemProbe probe,
        Func<Uri, CancellationToken, Task<bool>> reachable,
        Func<string, CancellationToken, Task<IReadOnlyList<string>>>? verifyFiles = null,
        bool deep = false,
        ExecutionProviderChoice? provider = null,
        CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(source);
        ArgumentNullException.ThrowIfNull(layout);
        ArgumentNullException.ThrowIfNull(probe);
        ArgumentNullException.ThrowIfNull(reachable);

        var findings = new List<DoctorFinding>();
        var folder = ModelDownloader.FolderFor(source, modelsRoot);
        var state = ModelDownloader.Inspect(folder, layout);
        var isPresent = state.IsComplete;

        findings.Add(ModelFinding(mode, isPresent, state));

        // When the model is missing the next checks are about the download,
        // not about files on disk (§11).
        var modelBytes = isPresent ? SizeOf(folder) : layout.ApproxDownloadBytes;

        findings.Add(DiskFinding(modelsRoot, !isPresent, modelBytes, probe));
        findings.Add(MemoryFinding(mode, modelBytes, isPresent, probe));

        if (!isPresent)
        {
            findings.Add(await SourceFinding(source, layout, reachable, ct).ConfigureAwait(false));
        }

        findings.Add(ProviderFinding(provider ?? OnnxRuntimeEnv.Current, probe));
        findings.Add(OutputFinding(outputFolder));

        if (deep && verifyFiles is not null && isPresent)
        {
            findings.Add(await IntegrityFinding(folder, verifyFiles, ct).ConfigureAwait(false));
        }

        return new DoctorReport(mode, findings);
    }

    /// <summary>
    /// Check 1. A missing model is not itself a failure — a generation
    /// downloads it.
    /// </summary>
    private static DoctorFinding ModelFinding(TtsMode mode, bool isPresent, ModelCompleteness state)
    {
        if (isPresent)
        {
            return new DoctorFinding(
                "Model", $"The {mode.DisplayName().ToLowerInvariant()} model is downloaded and ready.",
                DoctorSeverity.Ok);
        }

        var detail = state.Missing.Count == 0 && state.Partial.Count == 0
            ? "It has not been downloaded yet. It will download the first time you generate."
            : $"It is not complete ({state.Describe()}). The missing parts will download "
              + "the first time you generate.";

        return new DoctorFinding("Model", detail, DoctorSeverity.Ok);
    }

    /// <summary>Check 2. Measured on the volume holding the models folder (§3d).</summary>
    private static DoctorFinding DiskFinding(
        string modelsRoot, bool needsDownload, long modelBytes, ISystemProbe probe)
    {
        var free = probe.FreeSpaceBytes(modelsRoot);
        if (free is null)
        {
            return new DoctorFinding(
                "Disk space", "Could not work out how much room is free where models are kept.",
                DoctorSeverity.Warning);
        }

        var freeText = DownloadProgress.Bytes(free.Value);

        if (!needsDownload)
        {
            if (free.Value < OutputHeadroom)
            {
                return new DoctorFinding(
                    "Disk space",
                    $"Only {freeText} free. Generated audio is small, but a disk this full "
                    + "may not accept it. Free up some space.",
                    DoctorSeverity.Blocker);
            }

            return new DoctorFinding("Disk space", $"{freeText} free.", DoctorSeverity.Ok);
        }

        var needText = DownloadProgress.Bytes(modelBytes);

        if (free.Value < modelBytes)
        {
            var shortfall = DownloadProgress.Bytes(modelBytes - free.Value);
            return new DoctorFinding(
                "Disk space",
                $"{freeText} free, and the download needs about {needText}. "
                + $"Free up {shortfall}, or keep models on another drive in Settings.",
                DoctorSeverity.Blocker);
        }

        if (free.Value < modelBytes + DiskWarningHeadroom)
        {
            return new DoctorFinding(
                "Disk space",
                $"{freeText} free and the download needs about {needText}, which fits "
                + "but leaves the disk nearly full.",
                DoctorSeverity.Warning);
        }

        return new DoctorFinding(
            "Disk space", $"{freeText} free, and the download needs about {needText}.",
            DoctorSeverity.Ok);
    }

    /// <summary>
    /// Check 3. Warning only, never a blocker.
    /// </summary>
    /// <remarks>
    /// §11: it is a prediction about a run that has not started, the figure
    /// moves the moment another app quits, and a machine under pressure still
    /// finishes — only slowly. Blocking on it would refuse runs that would have
    /// worked.
    /// </remarks>
    private static DoctorFinding MemoryFinding(
        TtsMode mode, long modelBytes, bool isPresent, ISystemProbe probe)
    {
        var available = probe.AvailableMemoryBytes();
        if (available is null)
        {
            return new DoctorFinding(
                "Memory", "Could not work out how much memory is free.", DoctorSeverity.Warning);
        }

        var needed = (long)(modelBytes * MemoryWorkingFactor);
        var neededText = DownloadProgress.Bytes(needed);
        var availableText = DownloadProgress.Bytes(available.Value);
        var once = isPresent ? string.Empty : " once it has downloaded";

        if (available.Value >= needed)
        {
            return new DoctorFinding(
                "Memory",
                $"{availableText} available, and the {mode.DisplayName().ToLowerInvariant()} "
                + $"model wants about {neededText}.",
                DoctorSeverity.Ok);
        }

        return new DoctorFinding(
            "Memory",
            $"{availableText} available and the {mode.DisplayName().ToLowerInvariant()} model "
            + $"wants about {neededText}{once}. It will still run, but the machine may swap, "
            + "which makes generating slow and playback stutter. Long text needs more again. "
            + "Closing other apps first will help.",
            DoctorSeverity.Warning);
    }

    /// <summary>
    /// Check 4. Only when a download is required.
    /// </summary>
    /// <remarks>
    /// §11: a dead self-hosted server should be reported as a dead server
    /// rather than as a missing model. Asking when the files are already on
    /// disk would report a problem the user does not have — the app is
    /// offline-capable once they are there.
    /// </remarks>
    private static async Task<DoctorFinding> SourceFinding(
        ModelSource source,
        ModelLayout layout,
        Func<Uri, CancellationToken, Task<bool>> reachable,
        CancellationToken ct)
    {
        var probe = ProbeUriFor(source, layout);
        var where = source switch
        {
            ModelSource.Repo repo => repo.Id,
            ModelSource.BaseUrl url => url.Url.AbsoluteUri,
            _ => "the configured source",
        };

        var ok = await reachable(probe, ct).ConfigureAwait(false);

        return ok
            ? new DoctorFinding("Model source", $"{where} is answering.", DoctorSeverity.Ok)
            : new DoctorFinding(
                "Model source",
                $"{where} is not answering, so the model cannot be downloaded. "
                + "Check your connection, and the address in Settings.",
                DoctorSeverity.Blocker);
    }

    /// <summary>
    /// A URL that proves the source is alive.
    /// </summary>
    /// <remarks>
    /// Built from the export's own required-file list rather than assuming
    /// <c>config.json</c> at the top level: the preset-voice export keeps its
    /// config under <c>embeddings/</c>, so a hardcoded probe would report a
    /// perfectly good server as dead.
    /// </remarks>
    internal static Uri ProbeUriFor(ModelSource source, ModelLayout layout)
    {
        var first = layout.RequiredFiles.FirstOrDefault()?.RelativePath
                    ?? layout.Files.FirstOrDefault()?.RelativePath
                    ?? "config.json";

        return source switch
        {
            ModelSource.Repo repo => new Uri($"https://huggingface.co/{repo.Id}/resolve/main/{first}"),
            ModelSource.BaseUrl url => new Uri(
                url.Url.AbsoluteUri.TrimEnd('/') + "/" + first),
            _ => throw new ArgumentOutOfRangeException(nameof(source)),
        };
    }

    /// <summary>
    /// Check 5a. Which provider the speech model will use, and what is missing
    /// when it is not the fast one.
    /// </summary>
    /// <remarks>
    /// <b>Never a warning</b>, even on a machine that could be 3.7x faster.
    /// Doctor runs before every generation, and §11 is explicit that "a
    /// preflight the user notices on a healthy machine is a bug" — a machine
    /// running correctly on the CPU is healthy. It is a row in the report
    /// someone opens, which is where a speedup they did not know about is
    /// findable without being nagged about it on every run.
    /// </remarks>
    private static DoctorFinding ProviderFinding(
        ExecutionProviderChoice provider,
        ISystemProbe probe)
    {
        if (provider != ExecutionProviderChoice.Cpu)
        {
            return new DoctorFinding(
                "Acceleration",
                $"The speech model runs on {provider.Label()}. The final step, which turns "
                + "the result into sound, always runs on the CPU.",
                DoctorSeverity.Ok);
        }

        // "CPU on a machine with an NVIDIA card" is the case worth explaining.
        // "CPU on a machine without one" is simply the answer, and saying what
        // is missing there would be telling someone to buy hardware.
        return probe.HasNvidiaDriver() == true
            ? new DoctorFinding(
                "Acceleration",
                "The speech model runs on the CPU, though this machine has an NVIDIA "
                + "graphics card. CUDA measured 3.7x faster on long text. It needs the "
                + "CUDA build of Bunyi and NVIDIA's CUDA Toolkit.",
                DoctorSeverity.Ok)
            : new DoctorFinding(
                "Acceleration",
                "The speech model runs on the CPU, which is the supported configuration here.",
                DoctorSeverity.Ok);
    }

    /// <summary>Check 5. Blocker: there is nowhere to put the result.</summary>
    private static DoctorFinding OutputFinding(string folder)
    {
        try
        {
            Directory.CreateDirectory(folder);

            // Actually write, rather than inferring from attributes: a folder
            // can look writable and refuse, on a full disk or a read-only
            // mount.
            var probe = Path.Combine(folder, $".bunyi-write-test-{Guid.NewGuid():N}");
            File.WriteAllBytes(probe, [0]);
            File.Delete(probe);

            return new DoctorFinding("Output folder", $"Writable: {folder}", DoctorSeverity.Ok);
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            return new DoctorFinding(
                "Output folder",
                $"Cannot write to {folder}, so there is nowhere to save the audio. {ex.Message}",
                DoctorSeverity.Blocker);
        }
    }

    /// <summary>
    /// Check 6. On demand only.
    /// </summary>
    /// <remarks>
    /// §11: hashing gigabytes before every generation would be a worse problem
    /// than the one it detects. This is the check that catches a truncated or
    /// half-synced model — the failure that otherwise loads and speaks
    /// nonsense.
    /// </remarks>
    private static async Task<DoctorFinding> IntegrityFinding(
        string folder,
        Func<string, CancellationToken, Task<IReadOnlyList<string>>> verifyFiles,
        CancellationToken ct)
    {
        try
        {
            var bad = await verifyFiles(folder, ct).ConfigureAwait(false);

            if (bad.Count == 0)
            {
                return new DoctorFinding(
                    "Model files", "Every file matches the checksums your server published.",
                    DoctorSeverity.Ok);
            }

            return new DoctorFinding(
                "Model files",
                $"{bad.Count} file(s) do not match the checksums your server published "
                + $"({string.Join(", ", bad.Take(3))}). Delete the model in Settings and "
                + "download it again.",
                DoctorSeverity.Blocker);
        }
        catch (Exception ex) when (ex is IOException or HttpRequestException)
        {
            return new DoctorFinding(
                "Model files", $"Could not check the files against their checksums. {ex.Message}",
                DoctorSeverity.Warning);
        }
    }

    private static long SizeOf(string folder)
    {
        try
        {
            return Directory.Exists(folder)
                ? Directory.EnumerateFiles(folder, "*", SearchOption.AllDirectories)
                    .Sum(f => new FileInfo(f).Length)
                : 0;
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            return 0;
        }
    }
}
